### PR TITLE
basic framework for QEMU support in runtime-rs

### PR DESCRIPTION
**Pull request by pmores**
This PR adds basic boilerplate for QEMU driver in runtime-rs. It's obviously far from a full QEMU support but it does enable runtime-rs to start a very basic QEMU process (along with its supporting virtiofsd instance). It also includes (an embryo of) an alternative shim binary that drops containerd dependency and thus makes it fast and easy to execute shim proper, greatly enhancing dev cycle.

The main purpose of this WIP PR is to publish progress so far and give a common base to enable cooperation on QEMU support.

Fixes: #5817

### CODE REVIEW

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (Ji-Xinyou) [inline src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:31]: I think you should remove all the lines like `// println!(...)`

File: src/runtime-rs/crates/hypervisor/src/qemu/mod.rs
Comment (Ji-Xinyou) [inline src/runtime-rs/crates/hypervisor/src/qemu/mod.rs:17]: Maybe we need a hint on this commit that this is still a framework and dummy implementation like 👇🏻
```Rust
// Currently WIP
```

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (Ji-Xinyou) [inline src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:88]: I think here should be a `todo!()`

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (Ji-Xinyou) [inline src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:92]: Still, this is not a correct implementation, so it would better be left as a `todo!()`

Comment (Ji-Xinyou) [reply src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:92]: This suggestion IMO applies to some of the functions following.

File: src/runtime-rs/crates/shim-ctl/src/main.rs
Comment (Ji-Xinyou) [inline src/runtime-rs/crates/shim-ctl/src/main.rs:11]: I think this is a possible test tool for further hypervisor testing and development, thus it needs some refactories.

But the refactories could be left to the PRs in the future, I am just mentioning it. I think this part is fine for now.

**Review by Ji-Xinyou — CHANGES_REQUESTED** [review]
It is really nice to see that the QEMU support stepping forward! Thanks @pmores.

A few initial suggestions and you need to `cargo fmt && cargo clippy` haha.

**Comment by Ji-Xinyou** [comment]
You also need to relate this PR to an issue, I think there is one already :)

**Comment by pmores** [comment]
@Ji-Xinyou Thanks a lot for your time and a quick review, I really appreciate it! I'll push an updated version asap.

> You also need to relate this PR to an issue, I think there is one already :)

I wish this wasn't necessary... ;-) Yes, there is an issue for missing qemu support however I hesitate to reference that one as this PR hardly solves it. I think I'd better make up a new one.

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (pmores) [reply src/runtime-rs/crates/hypervisor/src/qemu/inner.rs]: True, should be fixed now.

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (pmores) [reply src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:78]: Good point, fixed.

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (pmores) [reply src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:83]: Fixed in all cases.

File: src/runtime-rs/crates/shim-ctl/src/main.rs
Comment (pmores) [reply src/runtime-rs/crates/shim-ctl/src/main.rs:9]: Yeah, all of this is inherently WIP, and this will be true also for at least a bunch of upcoming PRs since I'd really like to do this work iteratively, out in the open so that people can point out mistakes and wrong directions early.  So there will be some rough edges inevitably - in code that is meant to be replaced by a subsequent PR.

**Comment by pmores** [comment]
I pushed a new version that aims to address review requests. There's a _slight_ change in the final commit where I added an explicit creation of tokio runtime that imitates what the actual shim does.

File: src/runtime-rs/crates/shim-ctl/src/main.rs
Comment (Ji-Xinyou) [reply src/runtime-rs/crates/shim-ctl/src/main.rs:9]: I agree.

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (Ji-Xinyou) [inline src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:73]: Also this const should be moved up at the beginning part of the file I think.
The url's 3 and 1024 should also be made as consts.

File: src/runtime-rs/crates/shim-ctl/src/main.rs
Comment (Tim-0731-Hzt) [reply src/runtime-rs/crates/shim-ctl/src/main.rs:9]: I think It could be raised as another pr, as a testing tool

**Review by Tim-0731-Hzt — APPROVED** [review]

File: src/runtime-rs/crates/shim-ctl/src/main.rs
Comment (Tim-0731-Hzt) [reply src/runtime-rs/crates/shim-ctl/src/main.rs:9]: I think this could be raised as another pr, as a testing tool

Comment (Ji-Xinyou) [reply src/runtime-rs/crates/shim-ctl/src/main.rs:9]: IMO it's ok to be in the same PR, since this would work as a tool for testing.

File: src/runtime-rs/crates/shim-ctl/Cargo.toml
Comment (liubin) [inline src/runtime-rs/crates/shim-ctl/Cargo.toml:2]: It seems this is a new crate, can you add a README.md to describe this crate?

File: src/runtime-rs/crates/hypervisor/src/lib.rs
Comment (liubin) [inline src/runtime-rs/crates/hypervisor/src/lib.rs:32]: It's better to add a blank line after line 32.

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (liubin) [inline src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:46]: Use `info`/... to replace `println`

File: src/runtime-rs/crates/runtimes/virt_container/src/lib.rs
Comment (liubin) [inline src/runtime-rs/crates/runtimes/virt_container/src/lib.rs:179]: This line contains only spaces, so can delete the spaces.

File: src/runtime-rs/crates/runtimes/virt_container/src/lib.rs
Comment (liubin) [inline src/runtime-rs/crates/runtimes/virt_container/src/lib.rs:184]: Maybe using `#[tokio::test]` instead of `#[test]` can avoid manually calling `tokio::runtime`.

File: src/runtime-rs/crates/shim-ctl/src/main.rs
Comment (liubin) [inline src/runtime-rs/crates/shim-ctl/src/main.rs:27]: Delete this line.

File: src/runtime-rs/crates/shim-ctl/src/main.rs
Comment (liubin) [reply src/runtime-rs/crates/shim-ctl/src/main.rs:9]: @pmores if this is a test tool, it's better to not include this crate from `runtime-rs` crate: https://github.com/kata-containers/kata-containers/pull/5798/files#diff-50772485e77fcc2166bbc194075f08556b8c057fa21fc2be205a3b2f94ecba22R4

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (pmores) [reply src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:76]: Done. I will say though that the scheme string and the agent port aren't hypervisor specific so in order to have clean code we should really hoist them to a higher-level more widely accessible location (the runtime's agent crate [defines them independently](https://github.com/kata-containers/kata-containers/blob/7fabfb2cf09dabbe646509a197077cdaa696c686/src/runtime-rs/crates/agent/src/sock/mod.rs#L29) just as the dragonball driver [does](https://github.com/kata-containers/kata-containers/blob/7fabfb2cf09dabbe646509a197077cdaa696c686/src/runtime-rs/crates/hypervisor/src/dragonball/inner_hypervisor.rs#L83), in this case even directly in the function that uses it just like I did). Having multiple definitions for the same thing doesn't seem all that better after all.

Also, the CID constant will go away again immediately after I figure out how to use the infrastructure provided by PR #5691 (for some reason, qemu dislikes when I use it with `unable to set guest cid: Address already in use`).

File: src/runtime-rs/crates/shim-ctl/Cargo.toml
Comment (pmores) [reply src/runtime-rs/crates/shim-ctl/Cargo.toml:2]: Good point, added.

File: src/runtime-rs/crates/hypervisor/src/lib.rs
Comment (pmores) [reply src/runtime-rs/crates/hypervisor/src/lib.rs:32]: True, fixed.

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (pmores) [reply src/runtime-rs/crates/hypervisor/src/qemu/inner.rs]: Right, so I'm keeping this local for now. I'll only be able to use `info!()` once I figure out how to set up logging (to console, not file) in shim-ctl.

File: src/runtime-rs/crates/runtimes/virt_container/src/lib.rs
Comment (pmores) [reply src/runtime-rs/crates/runtimes/virt_container/src/lib.rs]: Fixed (I viewed the formatting from the source code point of view rather than the contents of the "file" so I placed it like a closing bracket but a fair point).

File: src/runtime-rs/crates/runtimes/virt_container/src/lib.rs
Comment (pmores) [reply src/runtime-rs/crates/runtimes/virt_container/src/lib.rs]: Good point, fixed, thanks!

File: src/runtime-rs/crates/shim-ctl/src/main.rs
Comment (pmores) [reply src/runtime-rs/crates/shim-ctl/src/main.rs]: Done, although it seems worth a mention that even the actual shim uses `println!()` at this level.

File: src/runtime-rs/crates/shim-ctl/src/main.rs
Comment (pmores) [reply src/runtime-rs/crates/shim-ctl/src/main.rs:9]: At this point it's more like a dev tool although it doesn't seem clear to me at this point what will evolve from it, if anything at all. I'm fine even keeping it local if necessary. My reasoning putting it where I did was that, it's basically the same thing as the actual shim itself, just with a different way of invocation, so I put it right next to the shim.

If there is a consensus where else to put it of course I'm happy to move it although then I'd prefer to do so in a separate PR as it would mean a fairly significant rework of the final commit of this one.

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (Ji-Xinyou) [reply src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:76]: I agree that may need some refactoring.

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (Ji-Xinyou) [inline src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:35]: I don't think these comments should be in a merged commit, you could add separate comments above this if you want to leave that as hints. But I think the comments in the form of code would be inappropriate.

File: src/runtime-rs/crates/shim-ctl/src/main.rs
Comment (Ji-Xinyou) [inline src/runtime-rs/crates/shim-ctl/src/main.rs:12]: remove the space between `new` and `(`

File: src/runtime-rs/crates/shim-ctl/src/main.rs
Comment (Ji-Xinyou) [inline src/runtime-rs/crates/shim-ctl/src/main.rs:29]: I think you should move const to the top lines of the file.

File: src/runtime-rs/crates/shim-ctl/src/main.rs
Comment (Ji-Xinyou) [inline src/runtime-rs/crates/shim-ctl/src/main.rs:37]: please remove the space

**Review by Ji-Xinyou — COMMENTED** [review]
Some more comments :)

**Comment by pmores** [comment]
Handled the latest batch of review requests by @Ji-Xinyou .

**Comment by Ji-Xinyou** [comment]
> Handled the latest batch of review requests by @Ji-Xinyou .

I'm good with the current code so far. You may mark resolve those reviews, which you have handled, as completed.

File: src/runtime-rs/crates/shim-ctl/src/main.rs
Comment (Tim-0731-Hzt) [inline src/runtime-rs/crates/shim-ctl/src/main.rs:18]: if you set `rootfs_mounts ` to empty, when the runtime is handling rootfs, I think it would fail.

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (c3d) [inline src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:48]: Why not `todo!()` here as well?

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (c3d) [inline src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:26]: Why not `todo!()` here? [This comment applies to `prepare_vm()`

File: src/runtime-rs/crates/hypervisor/src/qemu/mod.rs
Comment (c3d) [inline src/runtime-rs/crates/hypervisor/src/qemu/mod.rs:41]: There is a lot of repeated boilerplate code to forward a request to `inner`. I wonder if this would not be worth defining a macro.

File: src/runtime-rs/crates/shim-ctl/Cargo.toml
Comment (c3d) [inline src/runtime-rs/crates/shim-ctl/Cargo.toml:2]: I like the idea of adding a new tool for this, but maybe a different PR entirely?

File: src/runtime-rs/crates/shim-ctl/src/main.rs
Comment (pmores) [reply src/runtime-rs/crates/shim-ctl/src/main.rs]: Good point, fixed.

File: src/runtime-rs/crates/shim-ctl/src/main.rs
Comment (pmores) [reply src/runtime-rs/crates/shim-ctl/src/main.rs:19]: That's true, but it's also true of a vast majority of other configuration options since I can't do everything right away.  The current goal is to have just the simplest container started and stopped cleanly, then go back and gradually implement remaining features.  I don't feel comfortable enabling features before I'm ready to work on them.

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (pmores) [reply src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:48]: That's because this version of the qemu driver has nothing to do here, it's fine with nothing being done in `prepare_vm()`.

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (pmores) [reply src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:26]: That's because this version of the qemu driver has nothing to do here, it's fine with nothing being done in prepare_vm().

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (pmores) [reply src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:48]: True, `todo!()` would have been cleaner in this commit, however this is fixed by a subsequent commit in this very PR.

File: src/runtime-rs/crates/hypervisor/src/qemu/mod.rs
Comment (pmores) [reply src/runtime-rs/crates/hypervisor/src/qemu/mod.rs:41]: Makes sense, that could be an interesting exercise. For now I did like the dragonball driver does.

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (c3d) [reply src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:26]: Curious. The DragonBall variant does some non-trivial stuff, like setting up the VM path. Why would you not need that for `qemu`?

Comment (pmores) [reply src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:26]: I assume I will need that at some point, just not yet.  Frankly a more pressing part of `prepare_vm()` implementation at the moment is vsock CID allocation which is essential for a sandbox startup and so far only works for me if I hardcode the CID. I'm sure I'll get to the rest in due time. The immediate goal at the moment being just having a very simple container to launch, my next step is handling sandbox persistence.

File: src/runtime-rs/crates/hypervisor/src/qemu/mod.rs
Comment (Ji-Xinyou) [reply src/runtime-rs/crates/hypervisor/src/qemu/mod.rs:41]: > There is a lot of repeated boilerplate code to forward a request to `inner`. I wonder if this would not be worth defining a macro.

TBH, I once notice this and tries to work on this. But currently, the macro expansion order is not determined in Rust, therefore, we cannot achieve that if we still want to use the `async_trait` proc macro.

I roughly have some ideas on this and I will try to "macroize" this part of the code soon.

**Comment by studychao** [comment]
Could you raise an issue so that we could relate it to the issue which is a must for merging this into main :)?

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (studychao) [inline src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:72]: maybe #todo is needed here for future vsock related support work?

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (studychao) [inline src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:106]: I believe there is more capibilities for QEMU like `BlockDeviceSupport` or `BlockDeviceHotplugSupport`? 

Will those added later ?

**Comment by pmores** [comment]
> Could you raise an issue so that we could relate this PR to the issue which is a must for merging this into main :)?

Oh I believe I did that, didn't I? There's a 'Fixes: #5817' in the final commit of this series.

**Comment by pmores** [comment]
> > > Could you raise an issue so that we could relate this PR to the issue which is a must for merging this into main :)?
> > 
> > 
> > Oh I believe I did that, didn't I? There's a 'Fixes: #5817' in the final commit of this series.
> 
> Yes~~ But 5817 was not mentioned in the PR itself so the issue #5817 wasn't linked to this PR. But right now you have mentioned #5817 in this comment, so it's linked now fireworks
> 
> image

Ah okay, so it needs to be in both places. I'm sorry, I misunderstood.

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (pmores) [reply src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:72]: I'm actually working on this already as we speak ;-) I'm trying to employ the vsock code added by PR  #5691 to support dynamic allocation of CID, however qemu somehow disagrees with the result so far. Trying to figure it out.

Comment (studychao) [reply src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:72]: So maybe this work could be done in later PR and we could first add a `#todo` here?

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (pmores) [reply src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:106]: Yes, as I mentioned in another comment I'd rather not pretend supporting a feature until it's actually supported or I'm at least ready to work on it.  This code is currently correct from my standpoint since it rightly declares that this qemu driver supports no other storage than virtiofsd at the moment. This will obviously change in the future.

Comment (studychao) [reply src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:106]: ok~

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (pmores) [reply src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:72]: I'm not sure if you mean a todo as in a comment or a todo as in the macro.  I can sure add a comment but I think calling `todo!()` here would not be helpful as it would make it impossible to run and test the code. As is, this code is already able to launch a VM although with a limitation that there can be just a single such VM on a host - which is fine for this stage of development. Once the dynamic CID allocation work succeeds the limitation will be lifted.

File: src/runtime-rs/crates/shim-ctl/src/main.rs
Comment (pmores) [reply src/runtime-rs/crates/shim-ctl/src/main.rs:19]: Ah okay, having worked on this a bit more i can see your point now, sorry.  Apparently no container will start with empty `root_mounts` if I understand it correctly now.

Fixing this looks like a significant job though.  If anybody has any hints as to what to put into `rootfs_mounts`, that would be greatly appreciated.

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (studychao) [reply src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:72]: I mean the todo in the comment so that next time when we read through the code we could know that this part could be optimized and what to optimize :)

Example for the todo comments: https://github.com/kata-containers/kata-containers/blob/main/src/agent/src/netlink.rs#L519

File: src/runtime-rs/crates/shim-ctl/src/main.rs
Comment (Tim-0731-Hzt) [reply src/runtime-rs/crates/shim-ctl/src/main.rs:19]: My suggestion is to get rid of the `shim-ctl` from this pr this time and let's enable this in the future. What do you think?

File: src/runtime-rs/crates/hypervisor/src/qemu/inner.rs
Comment (pmores) [reply src/runtime-rs/crates/hypervisor/src/qemu/inner.rs:72]: Okay, I'll add the comment.

File: src/runtime-rs/crates/shim-ctl/src/main.rs
Comment (pmores) [reply src/runtime-rs/crates/shim-ctl/src/main.rs:19]: To be perfectly honest I'd appreciate if we can merge this soon as having it outstanding is an obstacle to my further developing qemu support.  I've been trying to continue development not to waste time but I'm often hitting the same lines of code that we've been polishing here.  When that happens I have to stash my changes, do a rebase & force push to change the PR and finally pop the stash which causes a merge conflict and a manual resolution.  This is tedious, time-consuming and error-prone (apart from the general risk of resulting in bad code there's a particular risk of losing the cosmetic changes again).

So I can break the shim-ctl commit out to a separate PR but I fear that then I'd have to do more weird stuff in my repo to continue development (I'd at least need to duplicate the commit) - or be blocked until both PRs are merged.  I agree that including the commit here is a border case but at the same time, it doesn't seem totally unreasonable so I'd lean towards keeping it here at this point.

Hopefully I'm making some sense. ;-)

File: src/runtime-rs/crates/shim-ctl/README.md
Comment (bergwolf) [inline src/runtime-rs/crates/shim-ctl/README.md:3]: Can we merge it into [kata-ctl](https://github.com/kata-containers/kata-containers/blob/d3ee8d9f1bcafe85991e98ea75bb9a941e8d97ef/src/tools/kata-ctl/README.md) so that we have a unified command line tool for the rust shim?

File: src/runtime-rs/crates/shim-ctl/src/main.rs
Comment (Tim-0731-Hzt) [reply src/runtime-rs/crates/shim-ctl/src/main.rs:19]: Yes, that makes sense to me. But I have another concern, `shim-ctl` will be compiled to another binary, is it possible to integrate this into `kata-ctl `? /cc @jodh-intel

File: src/runtime-rs/crates/shim-ctl/README.md
Comment (bergwolf) [reply src/runtime-rs/crates/shim-ctl/README.md:3]: Not required for this PR (as it has already been reviewed several rounds and is in good shape as is). But I'd like to get folks on the same page if we will merge shim-ctl into kata-ctl. /cc @jodh-intel

Comment (pmores) [reply src/runtime-rs/crates/shim-ctl/README.md:3]: I'm perfectly fine to do that.  I'd wait though to first see how this pans out. I'm still unsure if the approach taken in shim-ctl is even workable. Getting rid of the dependency on containerd is great and has enabled me to work on this so far but it also means that shim-ctl has to do some of the stuff normally done by containerd. I've no idea so far how workable this is - in the worst case it might mean duplicating of lot of containerd functionality in shim-ctl which might not be worth the trouble in the end.

So my suggestion would be to wait a bit until the approach is proven, then merge it into kata-ctl.

Comment (bergwolf) [reply src/runtime-rs/crates/shim-ctl/README.md:3]: Sounds good to me. Thanks @pmores !

**Review by studychao — APPROVED** [review]
Thanks, @pmores !

**Review by Tim-0731-Hzt — APPROVED** [review]
Thanks @pmores  LGTM

**Comment by Tim-0731-Hzt** [comment]
@pmores  The static check is failed, could you check that

**Comment by Tim-0731-Hzt** [comment]
ERROR: Document src/runtime-rs/crates/shim-ctl/README.md is not referenced

script:
  name: '/home/runner/work/kata-containers/kata-containers/src/github.com/kata-containers/tests/.ci/static-checks.sh'
  pid: 8684
  directory: '/home/runner/work/kata-containers/kata-containers/src/github.com/kata-containers/kata-containers'
  details: 'drwxr-xr-x 10 runner docker ? 4096 Dec  7 16:15 /home/runner/work/kata-containers/kata-containers/src/github.com/kata-containers/kata-containers'
failure:
  function: 'static_check_docs'
  file: '/home/runner/work/kata-containers/kata-containers/src/github.com/kata-containers/tests/.ci/static-checks.sh'
  line: 681
  name: '/home/runner/work/kata-containers/kata-containers/src/github.com/kata-containers/tests/.ci/static-checks.sh'

**Comment by Tim-0731-Hzt** [comment]
@pmores I just add support to start a container from the bundle https://github.com/kata-containers/kata-containers/pull/5877, and I test it with `containerd`. You can check that, and we can add this to `shim-ctl` in the future.

**Comment by studychao** [comment]
@pmores The error `ERROR: Document src/runtime-rs/crates/shim-ctl/README.md is not referenced` means that you should put this README into any other .md file as a reference. Please find a space for this README (probably under runtime-rs readme?)

**Comment by pmores** [comment]
@studychao Thanks for pointing this out, I've just pushed a new version that references the new README.

**Comment by pmores** [comment]
> @pmores I just add support to start a container from the bundle #5877, and I test it with `containerd`. You can check that, and we can use this for `shim-ctl` in the future.

That's an interesting idea, I'm taking a closer look.  It's a bit frustrating that detailed information about what exactly is done for root mounts before a runtime is invoked seems rather hard to come by.

**Comment by Tim-0731-Hzt** [comment]
> > @pmores I just add support to start a container from the bundle #5877, and I test it with `containerd`. You can check that, and we can use this for `shim-ctl` in the future.
> 
> That's an interesting idea, I'm taking a closer look. It's a bit frustrating that detailed information about what exactly is done for root mounts before a runtime is invoked seems rather hard to come by.

As I said before, it will include the information of snapshot directory. But if we pass a bundle to `containerd`, the rootfs mount will be empty. So, if we enable start container with the bundle in containerd, it could also work with `shim-ctl`.

**Review by bergwolf — APPROVED** [review]
lgtm!

**Pull request merged**
